File: mac/LRActionKit/LRActionKit/Action/Action.m
@import LRCommons;
@import PackageManagerKit;
@import ATPathSpec;

#import "Action.h"
#import "LRManifestLayer.h"
#import "LRActionVersion.h"
#import "LRActionManifest.h"
#import "LRAssetPackageConfiguration.h"
#import "LRActionKit-Swift.h"
#import "ActionKitSingleton.h"


@implementation Action {
    NSMutableArray *_errors;
    NSArray *_packageConfigurations;
    NSString *_fakeChangeExtension;
}

- (instancetype)initWithManifest:(NSDictionary *)manifest container:(id<ActionContainer>)container {
    if (self = [super initWithManifest:manifest errorSink:container]) {
        _container = container;
        [self initializeWithOptions];
    }
    return self;
}

- (void)initializeWithOptions {
    _identifier = [self.manifest[@"id"] copy] ?: @"";
    _name = [self.manifest[@"name"] copy] ?: _identifier;

    NSDictionary *knownTypes = @{
                                 @"filter": @{
                                         @"kind": @"filter",
                                         @"objc_class": [FilterRule class],
                                         @"objc_rowClass": NSClassFromString(@"FilterRuleRow"),
                                     },
                                 @"compile-file": @{
                                         @"kind": @"compiler",
                                         @"objc_class": [CompileFileRule class],
                                         @"objc_rowClass": NSClassFromString(@"CompileFileRuleRow"),
                                     },
                                 @"compile-folder": @{
                                         @"kind": @"postproc",
                                         @"objc_class": [CompileFolderRule class],
                                         @"objc_rowClass": NSClassFromString(@"FilterRuleRow"),
                                     },
                                 @"run-tests": @{
                                         @"kind": @"postproc",
                                         @"objc_class":    [RunTestsRule class],
                                         @"objc_rowClass": NSClassFromString(@"FilterRuleRow"),
                                     },
                                 @"custom-command": @{
                                         @"kind": @"postproc",
                                         @"objc_class": [CustomCommandRule class],
                                         @"objc_rowClass": NSClassFromString(@"CustomCommandRuleRow"),
                                     },
                                 @"user-script": @{
                                         @"kind": @"postproc",
                                         @"objc_class": [UserScriptRule class],
                                         @"objc_rowClass": NSClassFromString(@"UserScriptRuleRow"),
                                     },
                                 };

    NSDictionary *manifest = self.manifest;

    NSString *typeName = self.manifest[@"type"];
    if (typeName) {
        NSDictionary *defaultTypeManifest = knownTypes[typeName];

        NSMutableDictionary *mergedOptions = [NSMutableDictionary new];
        [mergedOptions addEntriesFromDictionary:defaultTypeManifest];
        [mergedOptions addEntriesFromDictionary:manifest];
        manifest = [mergedOptions copy];
    }

    _kind = LRActionKindFromString(manifest[@"kind"] ?: @"");

    _actionClass = manifest[@"objc_class"];
    _rowClass = manifest[@"objc_rowClass"];

    if (_identifier.length == 0)
        [self addErrorMessage:@"'id' attribute is required"];

    if (_kind == ActionKindUnknown)
        [self addErrorMessage:[NSString stringWithFormat:@"'kind' attribute is required and must be one of %@", LRValidActionKindStrings()]];

    LRPackageManager *packageManager = [ActionKitSingleton sharedActionKit].packageManager;
    NSArray *versionInfoLayers = [(self.manifest[@"versionInfo"] ?: @{}) arrayByMappingEntriesUsingBlock:^id(NSString *packageRefString, NSDictionary *info) {
        if ([packageRefString hasPrefix:@"__"])
            return nil;
        LRPackageReference *reference = [packageManager packageReferenceWithString:packageRefString];
        return [[LRManifestLayer alloc] initWithManifest:info requiredPackageReferences:@[reference] errorSink:self];
    }];

    _manifestLayers = [[self.manifest[@"info"] arrayByMappingElementsUsingBlock:^id(NSDictionary *info) {
        return [[LRManifestLayer alloc] initWithManifest:info errorSink:self];
    }] arrayByAddingObjectsFromArray:versionInfoLayers];

    NSMutableArray *packageConfigurations = [NSMutableArray new];
    NSArray *packageConfigurationManifests = self.manifest[@"packages"];
    if (packageConfigurationManifests && ![packageConfigurationManifests isKindOfClass:NSArray.class]) {
        [self addErrorMessage:@"Invalid type of 'packages' key"];
    } else {
        for (NSArray *packagesInfo in packageConfigurationManifests) {
            if (![packagesInfo isKindOfClass:NSArray.class])
                [self addErrorMessage:@"Every package configuration must be an array"];
            [packageConfigurations addObject:[[LRAssetPackageConfiguration alloc] initWithManifest:@{@"packages": packagesInfo} errorSink:self]];
        }
    }
    _packageConfigurations = [packageConfigurations copy];

    if (_packageConfigurations.count == 0)
        _primaryVersionSpace = nil;
    else {
        LRAssetPackageConfiguration *configuration = [_packageConfigurations firstObject];
        LRPackageReference *reference = [configuration.packageReferences firstObject];
        _primaryVersionSpace = reference.type.versionSpace;
    }

    NSString *inputPathSpecString = self.manifest[@"input"];
    if (inputPathSpecString) {
        _combinedIntrinsicInputPathSpec = [ATPathSpec pathSpecWithString:inputPathSpecString syntaxOptions:ATPathSpecSyntaxFlavorExtended];
    } else {
        _combinedIntrinsicInputPathSpec = [ATPathSpec emptyPathSpec];
    }

    NSString *outputSpecString = self.manifest[@"output"];

    // fake-change mode support is currently hard-coded to target compilers that produce CSS files
    // (everything else triggers a full page reload anyway)
    if ([outputSpecString isEqualToString:@"*.css"]) {
        _fakeChangeExtension = @"css";
    }
}

- (NSString *)description {
    return [NSString stringWithFormat:@"%@ '%@' (%@, %@)", LRStringFromActionKind(_kind), _identifier, NSStringFromClass(_actionClass), NSStringFromClass(_rowClass)];
}

- (NSString *)fakeChangeDestinationNameForSourceFile:(ProjectFile *)file {
    if (_fakeChangeExtension) {
        NSString *relativePath = file.relativePath;
        if (![[relativePath pathExtension] isEqualToString:_fakeChangeExtension]) {
            return [[relativePath stringByDeletingPathExtension] stringByAppendingPathExtension:_fakeChangeExtension];
        }
    }
    return nil;
}

@end

File: mac/LRActionKit/LRActionKit/Action/Action.swift
import Foundation
import SwiftyFoundation
import PackageManagerKit
import PiiVersionKit
import ATPathSpec

public class Action : LRManifestBasedObject {

    public let container: ActionContainer

    public let identifier: String
    public let name: String
    public var kind: ActionKind
    
    public let ruleType: Rule.Type
    public let rowClassName: String

    private let fakeChangeExtension: String?

    public let combinedIntrinsicInputPathSpec: ATPathSpec

    public private(set) var packageConfigurations: [LRAssetPackageConfiguration]!

    public private(set) var manifestLayers: [LRManifestLayer]!

    public var primaryVersionSpace: LRVersionSpace?
    
    public let compilableFileTag: Tag?
    public let taggers: [Tagger]
    public private(set) var derivers: [Deriver]!

    public init(manifest: [String: AnyObject], container: ActionContainer) {
        self.container = container
        identifier = manifest["id"] ~|||~ ""
        name = manifest["name"] ~|||~ identifier
        // if (_identifier.length == 0)
        //     [self addErrorMessage:@"'id' attribute is required"];

        // if (_kind == ActionKindUnknown)
        //     [self addErrorMessage:[NSString stringWithFormat:@"'kind' attribute is required and must be one of %@", LRValidActionKindStrings()]];

        let type = manifest["type"] ~|||~ ""
        switch type {
        case "filter":
            kind = .Filter
            ruleType = FilterRule.self
            rowClassName = "FilterRuleRow"
        case "compile-file":
            kind = .Compiler
            ruleType = CompileFileRule.self
            rowClassName = "CompileFileRuleRow"
        case "compile-folder":
            kind = .Postproc
            ruleType = CompileFolderRule.self
            rowClassName = "FilterRuleRow"
        case "run-tests":
            kind = .Postproc
            ruleType = RunTestsRule.self
            rowClassName = "FilterRuleRow"
        case "custom-command":
            kind = .Postproc
            ruleType = CustomCommandRule.self
            rowClassName = "CustomCommandRuleRow"
        case "user-script":
            kind = .Postproc
            ruleType = UserScriptRule.self
            rowClassName = "UserScriptRuleRow"
        default:
            fatalError("Missing or unknown action type")
        }

        if let inputPathSpecString: String = manifest["input"]~~~ {
            combinedIntrinsicInputPathSpec = ATPathSpec(string: inputPathSpecString, syntaxOptions:.FlavorExtended)
        } else {
            combinedIntrinsicInputPathSpec = ATPathSpec.emptyPathSpec()
        }

        if let outputSpecString: String = manifest["output"]~~~ {
            if (outputSpecString == "*.css") {
                fakeChangeExtension = "css"
            } else {
                fakeChangeExtension = nil
            }
        } else {
            fakeChangeExtension = nil
        }
        
        var taggers: [Tagger] = []
        if kind == .Compiler {
            compilableFileTag = Tag(name: "compilable.\(identifier)")
            taggers.append(FileSpecTagger(spec: combinedIntrinsicInputPathSpec, tag: compilableFileTag!))
        } else {
            compilableFileTag = nil
        }
        self.taggers = taggers

        super.init(manifest: manifest, errorSink: container)

        var derivers: [Deriver] = []
        if kind == .Compiler {
            derivers.append(CompileFileRuleDeriver(action: self))
        }
        self.derivers = derivers

        // Manifests
        let packageManager = ActionKitSingleton.sharedActionKit().packageManager
        let versionInfo = (manifest["versionInfo"] as? [String: AnyObject]) ?? [:]
        let versionInfoLayers = mapIf(versionInfo) { (packageRefString, info) -> LRManifestLayer? in
            // no idea what this check means
            if packageRefString.hasPrefix("__") {
                return nil
            }
            let reference = packageManager.packageReferenceWithString(packageRefString)
            return LRManifestLayer(manifest: info as! [String: AnyObject], requiredPackageReferences: [reference], errorSink: self)
        }

        let infoDictionaries = ArrayValue(manifest["info"]) { $0 as? [String: AnyObject] } ?? []
        manifestLayers = infoDictionaries.map { LRManifestLayer(manifest: $0, errorSink: self) } + versionInfoLayers

        let packageConfigurationManifests = ArrayValue(manifest["packages"]) { ArrayValue($0) { StringValue($0) } } ?? []
        packageConfigurations = packageConfigurationManifests.map { packagesManifest in
            return LRAssetPackageConfiguration(manifest: ["packages": packagesManifest], errorSink: self)
        }

        if packageConfigurations.isEmpty {
            primaryVersionSpace = nil
        } else {
            // wow, that's quite a chain
            primaryVersionSpace = packageConfigurations[0].packageReferences[0].type.versionSpace
        }
    }

    public override var description: String {
        return "\(kind) '\(identifier)'"
    }

    public func fakeChangeDestinationNameForSourceFile(file: ProjectFile) -> String? {
        if let fakeChangeExtension = fakeChangeExtension {
            let relativePath = file.relativePath
            if relativePath.pathExtension == fakeChangeExtension {
                return nil
            } else {
                return relativePath.stringByDeletingPathExtension.stringByAppendingPathExtension(fakeChangeExtension)
            }
        } else {
            return nil
        }
    }
    
    public func newRule(#contextAction: LRContextAction, memento: NSDictionary?) -> Rule {
        return ruleType(contextAction: contextAction, memento: memento)
    }

}
